### PR TITLE
Completing exercise for Module 3

### PR DESCRIPTION
**Pull request by ankanbanerjee22**
------------------
#### React JS Global Learning Program
-------------------
#### MODULE 3 (Exercise) Implement the following components

- **MovieTile**
This component should render a movie tile from the list of movies.
It should take properties to receive image url, movie name, release year, and a list of relevant genres. Alternatively, you can define one component property to take an object with all movie info. Additionally, the component should receive a callback property to capture click event.
image
-------------------

- **MovieDetails**
This component will render movie details when a movie is selected from the list (clicked). The details include movie poster image on the left and the rest of info on the right.
The component should take properties to receive image url, movie name, release year, rating, duration and a description. Alternatively, you can specify a single property that accepts an object with all movie info.
<br>

image
-------------------

- **SortControl**
This component will render a label "Sort by" and a select control to the right. Select should have the following options:
Release Date
Title
<br>

image
-------------------
- **Add stories for the components using story book**
<br>

### Stories
Counter
image

Movie Search Box
image

Movie Genre list selection
image

Movie sort control
image

Movie selection tile
image

Movie Description
image

- **Add test cases for the respective components**
image
-------------------

### CODE REVIEW

File: module_1/first-react-app/src/App.js [hunk 1]
```diff
@@ -26,6 +33,24 @@ function App() {
     alert("Searching movies with name: " + query);
   }
 
+  const handleTileClick = (movie) => {
+    //alert("trying to select movie: ", movie);
```
Comment (andreyBalabukha) [inline module_1/first-react-app/src/App.js]: Please remove comments in the PR

File: module_1/first-react-app/src/App.js [hunk 2]
```diff
@@ -26,6 +33,24 @@ function App() {
...
+    console.log(movie.movieName);
```
Comment (andreyBalabukha) [inline module_1/first-react-app/src/App.js]: I suppose we don't need console.log, please remove

File: module_1/first-react-app/src/MovieMetadata.json
Comment (andreyBalabukha) [inline module_1/first-react-app/src/MovieMetadata.json:1]: please use a camelCase notation for the file name. MovieMetadata is not a component. Only React components should be in Uppercase.

File: module_1/first-react-app/src/App.js [hunk 3]
```diff
@@ -43,10 +68,50 @@ function App() {
...
+              )}
+            </div>
+            <div className="row">
+              {/*  fetching from movie */}
+              {MovieMetadata.map((movie) => {
+                return (
+                  <div className="col s3 m3 l3 xxl" key={movie.imageUrl}>
+                    <Movie imageUrl={movie.imageUrl} movieName={movie.movieName}
+                      releaseYear={movie.releaseYear} genres={movie.genres}
+                      onTileClick={() => handleTileClick(movie)} >
```
Comment (andreyBalabukha) [inline module_1/first-react-app/src/App.js]: `onTitleClick` ?

File: module_1/first-react-app/src/components/Movie.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    const displayedMovieName = movieName ? movieName : UNKNOWN_MOVIE;
+    const displayedReleaseYear = releaseYear ? releaseYear : NOT_APPLICABLE;
+
+    return (
+        <>
+            <div className="row" >
+                <div className="col s12 movie-list">
+                    <div className="card large" onClick={onTileClick} >
+                        <div className="card-image movie-tile" >
+                            <img src={imageUrl} height='100%' alt='...' />
```
Comment (andreyBalabukha) [inline module_1/first-react-app/src/components/Movie.js]: please put the correct value for the `alt`

File: module_1/first-react-app/src/components/Movie.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+                        </div>
+                        <div className="card-content black movie-description">
+                            <span className="badge green darken-3 new bold-text" data-badge-caption="">{displayedReleaseYear}</span>
```
Comment (andreyBalabukha) [inline module_1/first-react-app/src/components/Movie.js:20]: Do we need that attr: `data-badge-caption=""`?

File: module_1/first-react-app/src/components/MovieDetails.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
+import React, {useState} from 'react';
+import StarRating from './StarRating';
+import '../css/MovieDetails.css'
+import { NOT_APPLICABLE, UNKNOWN_MOVIE } from '../literals';
+
+const MovieDetails = ({ imageUrl, movieName, releaseYear, rating, duration, description, genres, onClose }) => {
+
+    const trimmedDescription = description.length <= 400 ? description : description.slice(0, 400).trim() + '.......';
```
Comment (andreyBalabukha) [inline module_1/first-react-app/src/components/MovieDetails.js]: Why do we need `.trim()` here?

File: module_1/first-react-app/src/components/MovieDetails.js [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+                                    <div className='col s6 m6'>
+                                        <p className='red-text'> <strong>{displayedReleaseYear}</strong></p>
+                                    </div>
+                                    <div className='col s6 m6'>
+                                        <p className='red-text'><strong>{duration}</strong> </p>
+                                    </div>
+                                </div>
+                                <div> <StarRating rating={rating} /></div>
+                                <p className='grey-text' ><strong>{displayedGenre}</strong></p>
+                                <p><br></br></p>
```
Comment (andreyBalabukha) [inline module_1/first-react-app/src/components/MovieDetails.js]: `<p><br></br></p>` can it be made with the CSS?

File: module_1/first-react-app/src/components/MovieSortBy.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    
+    useEffect(() => {
+        window.M.AutoInit();
+    }, []);
+
+    return (
+        <>
+            <div className="input-field dropdown-container">
+                <span className="label" >Sort By : </span>
+                <select id="dropdown" className='select' value={sortBy} onChange={(e) => onSortChange(e.target.value)}>
```
Comment (andreyBalabukha) [inline module_1/first-react-app/src/components/MovieSortBy.js]: `className="select"`

File: module_1/first-react-app/src/components/MovieSortBy.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+const MovieSortBy = ({ sortBy, onSortChange }) => {
+    
+    useEffect(() => {
+        window.M.AutoInit();
+    }, []);
+
+    return (
+        <>
+            <div className="input-field dropdown-container">
+                <span className="label" >Sort By : </span>
```
Comment (andreyBalabukha) [inline module_1/first-react-app/src/components/MovieSortBy.js]: `<span className="label">` - the space deleted

File: module_1/first-react-app/src/components/StarRating.js [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+  const filledStars = Math.floor(rating);
+  const hasHalfStar = rating % 1 !== 0;
+
+  const stars = Array.from({ length: 10 }, (_, index) => {
+    if (index < filledStars) {
+      return <i key={index} className="material-icons yellow-text">star</i>;
+    } else if (hasHalfStar && index === filledStars) {
+      return <i key={index} className="material-icons yellow-text">star_half</i>;
+    } else {
+          return <i key={index} className="material-icons grey-text">star_outline</i>;
```
Comment (andreyBalabukha) [inline module_1/first-react-app/src/components/StarRating.js]: please fix the indentation

File: module_1/first-react-app/src/css/Movie.css [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
+.genre {
+    padding-top: 10px;
+    color: grey;
+    font-size: 1em;
+    overflow: hidden;
+}
+
+.movie-description {
+    height: 25% !important;
```
Comment (andreyBalabukha) [inline module_1/first-react-app/src/css/Movie.css:24]: Can we avoid `!important`?

File: module_1/first-react-app/src/css/Movie.css [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+.bold-text {
+    font-weight: 500 !important;
```
Comment (andreyBalabukha) [inline module_1/first-react-app/src/css/Movie.css:28]: Can we avoid `!important`?

**Review by andreyBalabukha — CHANGES_REQUESTED** [review]

File: module_1/first-react-app/src/components/Movie.js (hunk 5 above)
Comment (ankanbanerjee22) [reply module_1/first-react-app/src/components/Movie.js:20]: Yes, this material component has a default value of 'new', so kept it blank as we did not want that value!

File: module_1/first-react-app/src/App.js (hunk 1 above)
Comment (ankanbanerjee22) [reply module_1/first-react-app/src/App.js]: Removed comments!

File: module_1/first-react-app/src/App.js (hunk 2 above)
Comment (ankanbanerjee22) [reply module_1/first-react-app/src/App.js]: Removed console.log() statement

File: module_1/first-react-app/src/App.js (hunk 3 above)
Comment (ankanbanerjee22) [reply module_1/first-react-app/src/App.js]: Have renamed it to onClick()

File: module_1/first-react-app/src/components/Movie.js (hunk 4 above)
Comment (ankanbanerjee22) [reply module_1/first-react-app/src/components/Movie.js]: Have changed it to be set same as the movie name!

File: module_1/first-react-app/src/components/MovieSortBy.js (hunk 8 above)
Comment (ankanbanerjee22) [reply module_1/first-react-app/src/components/MovieSortBy.js]: corrected

File: module_1/first-react-app/src/components/MovieSortBy.js (hunk 9 above)
Comment (ankanbanerjee22) [reply module_1/first-react-app/src/components/MovieSortBy.js]: Corrected as per suggestion

File: module_1/first-react-app/src/components/StarRating.js (hunk 10 above)
Comment (ankanbanerjee22) [reply module_1/first-react-app/src/components/StarRating.js]: Fixed in commit: d4871b3879a408f1fcc8f54e9322ded924683aaf

File: module_1/first-react-app/src/css/Movie.css (hunk 12 above)
Comment (ankanbanerjee22) [reply module_1/first-react-app/src/css/Movie.css:28]: We have used this, as the style was being overridden by materialize css

File: module_1/first-react-app/src/css/Movie.css (hunk 11 above)
Comment (ankanbanerjee22) [reply module_1/first-react-app/src/css/Movie.css:24]: We have used this, as the style was being overridden by materialize css

File: module_1/first-react-app/src/MovieMetadata.json
Comment (ankanbanerjee22) [reply module_1/first-react-app/src/MovieMetadata.json:1]: Have changed the name to moviemetadata.json

File: module_1/first-react-app/src/components/MovieDetails.js (hunk 6 above)
Comment (ankanbanerjee22) [reply module_1/first-react-app/src/components/MovieDetails.js]: corrected in commit: d4871b3879a408f1fcc8f54e9322ded924683aaf

File: module_1/first-react-app/src/components/MovieDetails.js (hunk 7 above)
Comment (ankanbanerjee22) [reply module_1/first-react-app/src/components/MovieDetails.js]: Removed <br> in commit: d4871b3879a408f1fcc8f54e9322ded924683aaf

**Review by andreyBalabukha — APPROVED** [review]

**Pull request merged**